### DISCUSSION TITLE
Tasmota thermostat rule

### DISCUSSION THREAD

**Discussion by dowidopo** [arendst/Tasmota#19965] [Q&A]
Hello, please i have 16CH tasmota relay and need help with rules. On 1-15 relay have a Electric Thermal Actuator and on 16 relay is floor heating relay. I would like to make rule that will turn on the 16 relay if ANY of 1-15 switches is ON and will turn it OFF when EVERY one of relay 1-15 are off. I made it with rules that turn on or off that relay if reley 1-16 is going on or off but it is not good because if relays are on and one of my room is heated and turn off the relay then it turn off relay 16 too. I was trying make it with flows in Homey but that is not good enough and i have to make sure that relay is turning off every time when all relay are off. Thank you

**Comment by sfromis** [comment]
If your 16-relay board has an ESP32, I'd recommend using Berry instead of rules.

Still, here is a structure (not tested) which should be workable with rules:
```
rule1
 on Power1#state do Scale1 %value%,0,1,-1,1 endon
 on Power2#state do Scale1 %value%,0,1,-1,1 endon
 on Power3#state do Scale1 %value%,0,1,-1,1 endon
 on Power4#state do Scale1 %value%,0,1,-1,1 endon
 on var1%state do add2 %value% endon
 on var2#state==0 do Power16 off endon
Rule2
 on var2#state>0 do Power16 on endon
```
One rule for each relay (only 4 in example, just add more) to set var1 to -1 or 1 when state changes, and then add this to var2, which will be a count of relays on. When going to 0, turn off number 16. When >0, turn on. If you set one-shot on rule set number 2 it will not trigger every time, only when going from false to true first time.

I'm not sure how reliable this will be if changing multiple relay states at the same time.

**Reply by sfromis** [reply]
With recent Tasmota versions, another approach could be for each relay to have a rule like this:
`on Power2#state do var2 x%Power1%%Power2%%Power3%%Power4% endon`
And then compare var2 to `x0000` (etc) to see if all are off or not.

The `x` at the start is to ensure that it is not a numeric value, as those cannot contain 15 decimal digits, but text should be ok.

**Reply by dowidopo** [reply]
Thank you for reply but it doesnt work for me. If i set rules it dont do anything if i press any of 1-15 relays.

**Reply by sfromis** [reply]
Well, you need to understand how rules work, what I posted was merely "a structure (not tested) which should be workable".

One detail is that you need to enable rules - and in general watch the console to see what happens.

**Reply by dowidopo** [reply]
Yes i have rules enabled and in console it shows scale like in the rule but didnt turn on relay 16

**Reply by dowidopo** [reply]
I could make it with flows in homey but there i have second problem and it is mqtt. It can find that tasmota and it is showing 16CH but i can not control all of that channels. I can control only 1-10 and if i want to turn on channel 11 it will turn on channel 1 if i turn on channel 16 it turning on channel 6 ... if 14 it turn on 4CH.

**Reply by sfromis** [reply]
Only saying "didnt" is a waste of time. Show details if you want suggestions.

**Reply by dowidopo** [reply]
Hello, here is a copy from console. I dont know if there is some log if is i can send it but need help how to got it.



09:09:37.071 MQT: stat/tasmota_94740A/RESULT = {"POWER1":"OFF"}
09:09:37.073 MQT: stat/tasmota_94740A/POWER1 = OFF
09:09:37.114 RUL: POWER1#STATE performs "Scale1 0,0,1,-1,1"
09:09:37.123 MQT: stat/tasmota_94740A/RESULT = {"Scale1":"-1.000"}
09:09:38.516 MQT: tele/tasmota_94740A/STATE = {"Time":"2023-11-11T09:09:38","Uptime":"1T19:38:48","UptimeSec":157128,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":22,"MqttCount":1,"POWER1":"OFF","POWER2":"ON","POWER3":"OFF","POWER4":"OFF","POWER5":"OFF","POWER6":"OFF","POWER7":"OFF","POWER8":"OFF","POWER9":"OFF","POWER10":"OFF","POWER11":"OFF","POWER12":"OFF","POWER13":"OFF","POWER14":"OFF","POWER15":"OFF","POWER16":"OFF","Wifi":{"AP":1,"SSId":"Poor","BSSId":"50:EB:F6:87:5A:F8","Channel":3,"Mode":"11n","RSSI":36,"Signal":-82,"LinkCount":1,"Downtime":"0T00:00:05"}}
09:09:38.526 MQT: stat/tasmota_94740A/RESULT = {"POWER2":"ON"}
09:09:38.529 MQT: stat/tasmota_94740A/POWER2 = ON
09:09:38.590 RUL: POWER2#STATE performs "Scale1 1,0,1,-1,1"
09:09:38.597 MQT: stat/tasmota_94740A/RESULT = {"Scale1":"1.000"}
09:09:40.335 MQT: tele/tasmota_94740A/STATE = {"Time":"2023-11-11T09:09:40","Uptime":"1T19:38:50","UptimeSec":157130,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","POWER5":"OFF","POWER6":"OFF","POWER7":"OFF","POWER8":"OFF","POWER9":"OFF","POWER10":"OFF","POWER11":"OFF","POWER12":"OFF","POWER13":"OFF","POWER14":"OFF","POWER15":"OFF","POWER16":"OFF","Wifi":{"AP":1,"SSId":"Poor","BSSId":"50:EB:F6:87:5A:F8","Channel":3,"Mode":"11n","RSSI":34,"Signal":-83,"LinkCount":1,"Downtime":"0T00:00:05"}}
09:09:40.345 MQT: stat/tasmota_94740A/RESULT = {"POWER2":"OFF"}
09:09:40.349 MQT: stat/tasmota_94740A/POWER2 = OFF
09:09:40.390 RUL: POWER2#STATE performs "Scale1 0,0,1,-1,1"
09:09:40.397 MQT: stat/tasmota_94740A/RESULT = {"Scale1":"-1.000"}
09:09:41.207 MQT: tele/tasmota_94740A/STATE = {"Time":"2023-11-11T09:09:41","Uptime":"1T19:38:51","UptimeSec":157131,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":30,"MqttCount":1,"POWER1":"ON","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","POWER5":"OFF","POWER6":"OFF","POWER7":"OFF","POWER8":"OFF","POWER9":"OFF","POWER10":"OFF","POWER11":"OFF","POWER12":"OFF","POWER13":"OFF","POWER14":"OFF","POWER15":"OFF","POWER16":"OFF","Wifi":{"AP":1,"SSId":"Poor","BSSId":"50:EB:F6:87:5A:F8","Channel":3,"Mode":"11n","RSSI":40,"Signal":-80,"LinkCount":1,"Downtime":"0T00:00:05"}}
09:09:41.216 MQT: stat/tasmota_94740A/RESULT = {"POWER1":"ON"}
09:09:41.220 MQT: stat/tasmota_94740A/POWER1 = ON
09:09:41.293 RUL: POWER1#STATE performs "Scale1 1,0,1,-1,1"
09:09:41.303 MQT: stat/tasmota_94740A/RESULT = {"Scale1":"1.000"}
09:11:02.441 MQT: tele/tasmota_94740A/STATE = {"Time":"2023-11-11T09:11:02","Uptime":"1T19:40:12","UptimeSec":157212,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER1":"ON","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","POWER5":"OFF","POWER6":"OFF","POWER7":"OFF","POWER8":"OFF","POWER9":"OFF","POWER10":"OFF","POWER11":"OFF","POWER12":"OFF","POWER13":"OFF","POWER14":"OFF","POWER15":"OFF","POWER16":"OFF","Wifi":{"AP":1,"SSId":"Poor","BSSId":"50:EB:F6:87:5A:F8","Channel":3,"Mode":"11n","RSSI":40,"Signal":-80,"LinkCount":1,"Downtime":"0T00:00:05"}}
09:12:02.472 CMD: Rule1
09:12:02.480 MQT: stat/tasmota_94740A/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":274,"Free":237,"Rules":"on Power1#state do Scale1 %value%,0,1,-1,1 endon  on Power2#state do Scale1 %value%,0,1,-1,1 endon  on Power3#state do Scale1 %value%,0,1,-1,1 endon  on Power4#state do Scale1 %value%,0,1,-1,1 endon  on var1%state do add2 %value% endon  on var2#state==0 do Power16 off endon"}}
09:12:31.350 CMD: Rule2
09:12:31.358 MQT: stat/tasmota_94740A/RESULT = {"Rule2":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":35,"Free":476,"Rules":"on var2#state>0 do Power16 on endon"}}

**Reply by sfromis** [reply]
Yeah, as I suggested, you need to understand how rules work - to spot the typo: `var1%state` should've been `var1#state`

And what you posted is the log where it is visible what happens, and what rules you have.  :+1:

**Reply by dowidopo** [reply]
Thank you for reply i changed it and now it is working. Now i understand how it work but i can not turn off relays too fast because it need some time to make scale. Please and dont know how to repair mqtt or how can i do some changes and settings? Because my 16Ch relay is showing like 16Ch bud it is handling 1-8relays by mqtt with 10-16 buttom. So if i press relay 1 it turns on but it can not read that and if i hit 10 so relay 1 is going on and the mqtt can see it is on.

**Reply by sfromis** [reply]
Speed considerations is the reason for me posting the alternate approach in the very next post above. No rules needed to keep a counter updated.

**Reply by dowidopo** [reply]
Hello, so now it is working for me without mqtt good i did it with http and it can control all 16 relays with no problem. But i would do it with that second rule but i dont know how it would look like. How did you mean this

With recent Tasmota versions, another approach could be for each relay to have a rule like this:
on Power2#state do var2 x%Power1%%Power2%%Power3%%Power4% endon
And then compare var2 to x0000 (etc) to see if all are off or not.

Could you send me some sample like in first post?

Thank you

**Reply by sfromis** [reply]
> And then compare var2 to x0000 (etc) to see if all are off or not.

```
on var2#state=x0000 do Power16 off endon
on var2#state!=x0000 do Power16 on endon
```
As already mentioned, using one-shot mode for the 2nd of those rules is relevant, even if it does not hurt to power on while it already is on.

**Reply by dowidopo** [reply]
So 
Rule 1 
on Power1#state do var2 x%Power1%%Power2%%Power3%%Power4% endon
on Power2#state do var2 x%Power1%%Power2%%Power3%%Power4% endon
on Power3#state do var2 x%Power1%%Power2%%Power3%%Power4% endon
...

and 

Rule2
on var2#state=x0000 do Power16 off endon
on var2#state!=x0000 do Power16 on endon

?

**Reply by sfromis** [reply]
Yes, except for he detail about one-shot, which is mostly relevant to the "Power16 on" rule. And one-shot is for a whole rule set, not just one rule.
